src/shims/posix/android/foreign_items.rs

```diff
@@ -0,0 +1,28 @@
+use rustc_middle::mir;
+use rustc_span::Symbol;
+use rustc_target::spec::abi::Abi;
+
+use crate::*;
+use shims::foreign_items::EmulateByNameResult;
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn emulate_foreign_item_by_name(
+        &mut self,
+        link_name: Symbol,
+        abi: Abi,
+        args: &[OpTy<'tcx, Tag>],
+        dest: &PlaceTy<'tcx, Tag>,
+        ret: mir::BasicBlock,
+    ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
+        let this = self.eval_context_mut();
+
+        match &*link_name.as_str() {
+            _ => {
+                // By default we piggyback off the items emulated for Linux. For now this functions
+                // serves to make adding android-specific syscalls easy
+                return shims::posix::linux::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, abi, args, dest, ret);
```

Comment:
I am not very happy about this fallback. The linux functions are *not* prepared for android use, and we didn't check that their implementation is correct against their android specification. Android has its completely separate libc, and the API we are implementing here is basically that of libc.

So, I am not convinced that we want to fall back like this here; to me this seems to carry a huge risk of accidentally using the wrong shim implementation.

Reply:
That makes sense. I originally went with this approach since Android *is* Linux-based, but you're right that if we're implementing the *libc* API, we should probably ensure these are separate.

What would you suggest, just copying the Linux code here?

Reply:
I would suggest copying enough of the Linux code to get `hello.rs` working on Android -- and leaving everything else to a future PR.